MNT blackify source code and add pre-commit

akash-suresh opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn-contrib/imbalanced-learn/blob/master/CONTRIBUTING.md#contributing-pull-requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Addressing https://github.com/scikit-learn-contrib/imbalanced-learn/issues/684

#### What does this implement/fix? Explain your changes.
Integrating black into the codebase, to keep the code format consistent. 

- [x] Integrate black
- [x] Run black over all files
- [x] Add black into precommit hook

#### Any other comments?

Open questions - 
1. Which `requirements` file should the `black` dependency be added to?
2. `line-length` for black is currently set as `79`. Is that alright?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

pep8speaks commented:
Hello @akash-suresh! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:































































* In the file [`imblearn/keras/_generator.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/keras/_generator.py):

> [Line 176:40](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/keras/_generator.py#L176): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 182:40](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/keras/_generator.py#L182): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 191:44](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/keras/_generator.py#L191): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'



* In the file [`imblearn/over_sampling/_smote.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/over_sampling/_smote.py):

> [Line 978:66](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/over_sampling/_smote.py#L978): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'





* In the file [`imblearn/tensorflow/_generator.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tensorflow/_generator.py):

> [Line 84:56](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tensorflow/_generator.py#L84): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 85:56](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tensorflow/_generator.py#L85): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 92:53](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tensorflow/_generator.py#L92): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'







* In the file [`imblearn/tests/test_pipeline.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tests/test_pipeline.py):

> [Line 1327:80](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tests/test_pipeline.py#L1327): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 1328:80](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tests/test_pipeline.py#L1328): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 1334:80](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/tests/test_pipeline.py#L1334): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)







* In the file [`imblearn/under_sampling/_prototype_selection/_nearmiss.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/under_sampling/_prototype_selection/_nearmiss.py):

> [Line 143:64](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/under_sampling/_prototype_selection/_nearmiss.py#L143): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 290:80](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/under_sampling/_prototype_selection/_nearmiss.py#L290): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 79 characters)







* In the file [`imblearn/under_sampling/_prototype_selection/tests/test_repeated_edited_nearest_neighbours.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/under_sampling/_prototype_selection/tests/test_repeated_edited_nearest_neighbours.py):

> [Line 367:40](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/under_sampling/_prototype_selection/tests/test_repeated_edited_nearest_neighbours.py#L367): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','



* In the file [`imblearn/utils/estimator_checks.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/utils/estimator_checks.py):

> [Line 402:67](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/utils/estimator_checks.py#L402): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','





* In the file [`imblearn/utils/tests/test_validation.py`](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/utils/tests/test_validation.py):

> [Line 349:45](https://github.com/scikit-learn-contrib/imbalanced-learn/blob/66d4e081d0e54c2fe871ce58984fb3043afdf2a9/imblearn/utils/tests/test_validation.py#L349): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

akash-suresh commented:
@chkoar @glemaitre `black` will always add spaces around `:`. As per `black` we will need to ignore the `E203` error thrown. Reference - https://github.com/psf/black/issues/315#issuecomment-395457972

glemaitre commented:
So let's go with 88 characters. We need to change the code in `build_tools/circle/linting.sh`.

glemaitre commented:
We can put the rules to ignore inside the `setup.cfg` like in pandas: v

akash-suresh commented:
> So let's go with 88 characters. We need to change the code in `build_tools/circle/linting.sh`.

Do you mean something like this 
```
git diff --unified=0 $COMMIT_RANGE -- $files | flake8 --diff --max-line-length=88 --show-source $options
```

akash-suresh commented:
@glemaitre How does the `pep8speaks` get generated? I am not able to address the errors thrown on those.

glemaitre commented:
We might need to add a config file: https://github.com/pandas-dev/pandas/blob/master/.pep8speaks.yml

We can add that we only check the patch and tune to the 88 characters.

akash-suresh commented:
@glemaitre I added the `.pep8speaks.yml`. [https://github.com/scikit-learn-contrib/imbalanced-learn/pull/722/files#diff-d468b3fa9b617bf1f88d5bc0a98969b2]

The pep8speaks analysis didn't get updated.

akash-suresh commented:
> @glemaitre I added the `.pep8speaks.yml`. https://github.com/scikit-learn-contrib/imbalanced-learn/pull/722/files#diff-d468b3fa9b617bf1f88d5bc0a98969b2
> 
> The pep8speaks analysis didn't get updated.

^ @glemaitre wondering if you found some free time, to take a look at this. 😄

chkoar commented:
A badge in the README?
![](https://camo.githubusercontent.com/b243a4527ef0973b38d54728651624816b6d7844/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f636f64652532307374796c652d626c61636b2d3030303030302e7376673f7374796c653d666c61742d737175617265)

akash-suresh commented:
@glemaitre @chkoar How do I fix the errors thrown by `pep8speaks`?

chkoar commented on imblearn/under_sampling/_prototype_selection/_nearmiss.py:
```diff
@@ -263,10 +263,13 @@ def _fit_resample(self, X, y):
 
         return _safe_indexing(X, idx_under), _safe_indexing(y, idx_under)
 
+    # fmt: off
```
Why this?

chkoar commented:
In general we should let `black` to do its job. `It is known that flake8` is not following strictly `pep8` it introduces some errors according its opinions. I pushed in master directly a .pep8speask.yml. This pr should be updated.

chkoar commented:
@akash-suresh I made some changes to your fork but pep8speaks does not seem to respond. Even if diff_only is False. I am may missing something.

chkoar commented:
@akash-suresh pep8speks got updated. You could continue solving its issues.

akash-suresh commented:
Sure @chkoar! Was a little busy with other things. Will add the pre-commit hooks, and update the PR.

akash-suresh commented:
Hey @chkoar. I have added the pre-commit hooks. 

I added the dependency for `pre-commit` in `requirements.txt`, should I add it in `requirements.optional.txt` and `requirements.test.txt` as well?

akash-suresh commented:
@chkoar Can we close this?

chkoar commented:
> I added the dependency for pre-commit in requirements.txt, should I add it in requirements.optional.txt and requirements.test.txt as well?

Do we use the requirement files somewhere? To be honest I thing that `.pre-commit-config.yaml` is nice to have but mandatory, in the sense that someone might not want to use it. Do we have documented that we use black?

glemaitre commented:
I think it would be best to add it in the config. We should add that we use black in the contribution guide. We should `git grep` for `flake8` since we will change to `black` then.

akash-suresh commented:
> > I added the dependency for pre-commit in requirements.txt, should I add it in requirements.optional.txt and requirements.test.txt as well?
> 
> Do we use the requirement files somewhere? To be honest I thing that `.pre-commit-config.yaml` is nice to have but mandatory, in the sense that someone might not want to use it. Do we have documented that we use black?

They can skip using the pre-commit hook by adding a `--no-verify` flag while committing.
Example - 
```
git commit -m "commit message" --no-verify
```

hayesall commented on imblearn/over_sampling/_smote.py:
```diff
@@ -1306,9 +1267,9 @@ def _fit_resample(self, X, y):
-                nns = self.nn_k_.kneighbors(
-                    X_cluster_class, return_distance=False
-                )[:, 1:]
+                nns = self.nn_k_.kneighbors(X_cluster_class, return_distance=False)[
+                    :, 1:
+                ]
```
Is this spilling over the max line length?

This looks like a case where `black` misses eliminating newlines.

hayesall commented on imblearn/pipeline.py:
```diff
@@ -149,8 +149,7 @@ def _validate_steps(self):
                 raise TypeError(
-                    "All intermediate steps of the chain should not be"
-                    " Pipelines"
+                    "All intermediate steps of the chain should not be" " Pipelines"
```
This should be switched to: `"All intermediate steps of the chain should not be Pipelines"`

hayesall commented on imblearn/tests/test_pipeline.py:
```diff
@@ -204,17 +204,14 @@ def test_pipeline_init():
     error_regex = (
-        "Last step of Pipeline should implement fit or be the "
-        "string 'passthrough'"
+        "Last step of Pipeline should implement fit or be the " "string 'passthrough'"
```
Another weird string concatenation case.

hayesall commented on imblearn/tests/test_pipeline.py:
```diff
@@ -1259,57 +1225,83 @@ def test_score_samples_on_pipeline_without_score_samples():
         AttributeError,
-        match="'LogisticRegression' object has no attribute "
-        "'score_samples'",
+        match="'LogisticRegression' object has no attribute " "'score_samples'",
```
Weird string concatenation.

hayesall commented on imblearn/tests/test_pipeline.py:
```diff
@@ -1259,57 +1225,83 @@ def test_score_samples_on_pipeline_without_score_samples():
-    with pytest.raises(ValueError, match="Pipeline.fit does not accept "
-                                         "the sample_weight parameter"):
+    with pytest.raises(
+        ValueError, match="Pipeline.fit does not accept " "the sample_weight parameter",
```
Weird string concatenation.

hayesall commented on imblearn/under_sampling/_prototype_selection/_nearmiss.py:
```diff
@@ -198,8 +193,7 @@ def _validate_estimator(self):
             raise ValueError(
-                "Parameter `version` must be 1, 2 or 3, got"
-                " {}".format(self.version)
+                "Parameter `version` must be 1, 2 or 3, got" " {}".format(self.version)
```
Weird string concatenation.

hayesall commented on imblearn/utils/_validation.py:
```diff
@@ -167,12 +164,9 @@ def _sampling_strategy_majority(y, sampling_type):
         raise ValueError(
-            "'sampling_strategy'='majority' cannot be used with"
-            " over-sampler."
+            "'sampling_strategy'='majority' cannot be used with" " over-sampler."
```
Weird string concatenation.

hayesall commented on imblearn/utils/_validation.py:
```diff
@@ -426,8 +402,7 @@ def _sampling_strategy_float(sampling_strategy, y, sampling_type):
         raise ValueError(
-            "'clean-sampling' methods do let the user "
-            "specify the sampling ratio."
+            "'clean-sampling' methods do let the user " "specify the sampling ratio."
```
Weird string concatenation.

glemaitre commented:
This would be cool to resolve the conflict and answer the suggestions to have it before the next release.
